crypto/ec/ec_lib.c

```diff
@@ -284,15 +284,22 @@ int EC_GROUP_set_generator(EC_GROUP *group, const EC_POINT *generator,
     if (order != NULL) {
         if (!BN_copy(group->order, order))
             return 0;
-    } else
+    } else {
         BN_zero(group->order);
+    }
 
+    /* The cofactor is an optional field, so it should be able to be NULL. */
     if (cofactor != NULL) {
-        if (!BN_copy(group->cofactor, cofactor))
+        if (group->cofactor == NULL) {
+            if ((group->cofactor = BN_dup(cofactor)) == NULL)
+                return 0;
+        } else if (!BN_copy(group->cofactor, cofactor)) {
             return 0;
-    } else
-        BN_zero(group->cofactor);
-
+        }
+    } else {
+        BN_free(group->cofactor);
+        group->cofactor = NULL;
+    }
```

Comment:
In that case the group compare cant just check for NULL.. Will revert but fix that issue also.

Reply:
> In that case the group compare cant just check for NULL.. Will revert but fix that issue also.

I am not sure I understand what you mean:
`EC_GROUP_cmp()` was not checking for `NULL` cofactors, it was using `BN_cmp(ac, bc)`.

> The problem with this is instead of it being NULL it could be be used as a zero value (which is not ideal :))

I agree it feels cleaner to have a `NULL` cofactor when it is undefined, instead of having it with a value of `0`, but we have to make sure that the semantics for existing applications and engines does not change!